### DISCUSSION TITLE
Can't get ST7789 working under Universal Display Driver

### DISCUSSION THREAD

**Discussion by Sum-Guy** [arendst/Tasmota#17569] [Q&A]
Now that I have the file system working on this ESP8266 nodemcu, I've uploaded this file:

Tasmota/tasmota/displaydesc/ST7789_display.ini 

But I first renamed it to display.ini before uploading it.  According to what I see in the file, the width and height is correct (240 x 240).

I have an ST7789 connected and configured.  Before I got the file system working the display was set to displaymodel 12 and I could display text in mode 0 and the splash was working and it displays "ST7789 TFT!" in the middle of the screen during device boot.  I have set GPIO16 to Option A3.

With the display ini file uploaded and the device re-booted, I see the splash text for a second or two before it goes away.

Console display command says 

{"Display":{"Model":17,"Type":0,"Width":800,"Height":600,"Mode":2,"Dimmer":10,"Size":1,"Font":1,"Rotate":0,"Invert":0,"Refresh":2,"Cols":[16,8],"Rows":2}}

Where it's getting 800 x 600 from, I don't know.  

I don't know how to read the rest of that line to know if it's following the display.ini file or not.  When I hard reset the device, the splash text comes up, but then goes blank.  The power toggle is ON.

If I change the displaymode to 0, I see nothing in response to displaytext commands.  Displaymode 1 or 2 or 3 likewise do nothing.

So what am I doing wrong now?

**Comment by gemu2015** [comment]
i assume you mixed up the old and the new driver. normally you would only compile udisplay.
also in the Tasmota GPIO GUI select spi_dc not ST7789 dc because this enables the old driver if you have both.

**Reply by Sum-Guy** [reply]
I have changed ST7789_DC to SPI_DC in the config, saved the config, and reset the device.  I no longer see the splash text.  But still the display does not work.  Nothing prints in textmode 0.

Do I have to compile tasmota without using the display drivers? But I still must have 

#define USE_UNIVERSAL_DISPLAY

in user config override?

In other words, what do I do with all the "#define use_display_what-ever lines when I am using the universal display?

**Reply by gemu2015** [reply]
you need at least these defines
#define USE_SPI
#define USE_DISPLAY
#define SHOW_SPLASH
#define USE_UNIVERSAL_DISPLAY

then define in Tasmota GPIO GUI

SPI_CLK
SPI_MOSI
SPI_DC
Backlight
Display_Rst
Option_A3

in console you should see init message from display =>

DSP: Configured display 'ST7789'

cmd display should show:
 
RESULT = {"Display":{"Model":17,"Type":0,"Width":240,"Height":240,"Mode":0,"Dimmer":60,"Size":1,"Font":1,"Rotate":0,"Invert":0,"Refresh":2,"Cols":[16,8],"Rows":2}}

**Reply by Sum-Guy** [reply]
Does backlight need to be configured?  I am not using it, but ok, I will waste another GPIO pin.

Here is how the module is now configured:

GPIO0 - backlight
GPIO2 - DHT11
GPIO4 - Display RST
GPIO5 - SPI DC
GPIO12 - SPI MISO
GPIO13 - SPI MOSI
GPIO14 - SPI CLK
GPIO16 - Option A3

When booting, this is what I see on serial port:

HDW: ESP8266EX
UFS: FlashFS mounted with 1984 kB free
CFG: Loaded from File, Count 32
QPC: Count 1
RSL: RESULT = {"POWER":"ON"}
RSL: POWER = ON
DSP: Configured display '@'
Project tasmota - Tasmota Version 12.1.1(TasmoCompiler-esp82664M)-2_7_4_9(2023-01-01T01:17:56)

Display command in console gives same info as I posted in the first post.  Display.ini is present in the file system.

Display is still not functional.  No splash text.

------------------------

I have 71 lines in user_config_override.h that start with "#define "

Some of them are duplicated - is that ok?  Or does that cause problems?

Yes, all of these are present:

#define USE_SPI
#define USE_DISPLAY
#define SHOW_SPLASH
#define USE_UNIVERSAL_DISPLAY

I also have 

#define USE_MULTI_DISPLAY

I have 19 lines that are

#define USE_DISPLAY_(something)

13 of those are a specific hardware chip

But like I said, I see there are some duplicates in USE_DISPLAY, is that a problem?

In general in the user config override, if there are duplicate #define lines, is that a problem for the compiler?

*Should I use some other way to flash the chip?  Is the tasmotizer trusted / ok or not?*

**Reply by Sum-Guy** [reply]
This is the display.ini file present on the device:

:H,ST7789,240,240,16,SPI,1,*,*,*,*,*,*,*,40
:S,2,1,3,0,80,30
:I
01,A0
11,A0
3A,81,55
36,81,00
21,80
13,80
29,A0
:o,28
:O,29
:A,2A,2B,2C
:R,36
:0,C0,00,50,00
:1,A0,50,00,01
:2,00,00,00,02
:3,60,00,00,03
:i,20,21
:TI2,38,32,23
#

I believe that each line is terminated with (CR)(LF)

**Reply by gemu2015** [reply]
you may have duplicate defines, should not matter.

the file display.ini should only have linefeed endings, not CR,LF

DSP: Configured display '@' indicates that something  is wrong with reading display.ini

**Reply by Sum-Guy** [reply]
When looking to download the file

Tasmota/tasmota/displaydesc/ST7789_display.ini

from github, there was no "download" link that I could find.  Every time I click on the file it opens the file in a browser window.

So I copied the text and pasted it in a windows notepad file.  That is where the CR/LF came from.   If there is a way to DOWNLOAD these files from github then it needs to be explained to me.

I have loaded the file into notepad++  and converted it into unix style (LF) and then uploaded it back to the ESP8266 tasmota file system.

I have reset the device.  Display still does not work.

I have then deleted the file in tasmota and used the create edit new file and pasted the contents into the tasmota web gui and saved the file.  I have downloaded that file back to my PC and checked it and the lines are terminated with LF only.

I reset the device and still I get nothing on the display.

DSP: Configured display '@'

Again I ask - is tasmotizer flash method ok or should I reflash the device using another method?

What else could be wrong?

**Reply by gemu2015** [reply]
i use that display with 2 devices but with esp32. however i also use an ili9341 with wemos esp8266 and have no issues at all.
try setting debug to "MORE" and check what is printed.
there is an extended serial debug mode in the driver itself where you have to uncomment in "uDisplay.cpp"
//#define UDSP_DEBUG

**Reply by Sum-Guy** [reply]
I've set seriallog to 4.  Is that sufficient?  Or do I need to add

#define UDSP_DEBUG

And re-compile?

With seriallog set to 4, and doing a restart, this is what I get (removing a few wifi-related lines):

00:00:00.001 HDW: ESP8266EX
00:00:00.012 UFS: FlashFS mounted with 1984 kB free
00:00:00.075 CFG: Loaded from File, Count 44
00:00:00.081 QPC: Count 1
00:00:00.083 CFG: CR 368/699, Busy 0
00:00:00.090 ROT: Mode 1
00:00:00.096 DSP: File descriptor used
00:00:00.100 SRC: Display
00:00:00.103 RSL: RESULT = {"POWER":"ON"}
00:00:00.104 RSL: POWER = ON
00:00:00.105 DSP: Configured display '@'
00:00:00.114 SRC: Restart
00:00:00.117 Project tasmota - Tasmota Version 12.1.1(TasmoCompiler-esp82664M)-2_7_4_9(2023-01-01T01:17:56)
00:00:00.118 DHT: (v6) Sensors found 1
00:00:01.802 WIF: Connected
00:00:02.055 HTP: Web server active on tasmota-9DF5B7-5559 with IP address 192.168.2.240
00:00:02.401 WIF: Sending Gratuitous ARP
00:00:02.402 NTP: Sync time...
00:00:02.517 RTC: UTC 2023-01-02T23:26:20, DST 2023-03-26T02:00:00, STD 2023-10-29T03:00:00
00:26:20.000 RTC: Synced by NTP
00:26:20.029 DHT: Pin2 read 240015033C
00:26:21.276 RSL: INFO1 = {"Info1":{"Module":"Generic","Version":"12.1.1(TasmoCompiler-esp82664M)","FallbackTopic":"cmnd/DVES_9DF5B7_fb/","GroupTopic":"cmnd/tasmotas/"}}
00:26:21.282 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-9DF5B7-5559","IPAddress":"192.168.2.240"}}
00:26:21.293 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":18}}
00:26:21.301 RSL: RESULT = {"POWER":"ON"}
00:26:21.304 RSL: POWER = ON
00:26:21.309 TFS: File 'autoexec.bat' not found
00:26:24.425 QPC: Reset
00:26:24.455 DHT: Pin2 read 230015033B
00:26:25.436 RSL: SENSOR = {"Time":"2023-01-03T00:26:25","DHT11":{"Temperature":21.3,"Humidity":35.0,"DewPoint":5.2},"TempUnit":"C"}
00:26:26.385 APP: Boot Count 18
00:26:27.191 CFG: Saved to flash at 3FB, Count 45, Bytes 4096
00:26:28.426 DHT: Pin2 read 240015033C
00:26:30.427 RSL: config = {"ip":"192.168.2.240","dn":"Tasmota","fn":["Tasmota",null,null,null,null,null,null,null],"hn":"tasmota-9DF5B7-5559","mac":"BCDDC29DF5B7","md":"Generic","ty":0,"if":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"12.1.1","t":"tasmota_9DF5B7","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[1,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null],"btn":[0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":0,"lt_st":0,"sho":[0,0,0,0],"sht":[[0,0,0],[0,0,0],[0,0,0],[0,0,0]],"ver":1}
00:26:30.477 RSL: sensors = {"sn":{"Time":"2023-01-03T00:26:30","DHT11":{"Temperature":21.3,"Humidity":36.0,"DewPoint":5.6},"TempUnit":"C"},"ver":1}
00:26:32.407 DHT: Pin2 read 240015033C
00:26:36.418 DHT: Pin2 read 240015033C

Would there be any additional info if UDSP_DEBUG was used?

I take it that the important part is this:

00:00:00.075 CFG: Loaded from File, Count 44
00:00:00.081 QPC: Count 1
00:00:00.083 CFG: CR 368/699, Busy 0
00:00:00.090 ROT: Mode 1
00:00:00.096 DSP: File descriptor used
00:00:00.100 SRC: Display
00:00:00.103 RSL: RESULT = {"POWER":"ON"}
00:00:00.104 RSL: POWER = ON
00:00:00.105 DSP: Configured display '@'

I don't see anything useful there. ?

> try setting debug to "MORE"

Where do I do that?  debug is not a valid console command...

**Reply by sfromis** [reply]
Debug "MORE" is log level 4, in `Weblog`, `Seriallog` or whatever type of logging you use. `#define UDSP_DEBUG` would be in addition to the log level.

That `DSP: Configured display '@'` indeed suggests that something went wrong with the `display.ini` file. If you have passed it through some editor or copy/paste (especially on Windows), there is a risk that the file was mangled by that, even without you changing anything.

I'd suggest downloading it by going to:
https://github.com/arendst/Tasmota/blob/development/tasmota/displaydesc/ST7789_display.ini
and then download it by right-clicking on the "Raw" button, and select "Save link as..." (details may vary depending on your browser).

You should then be able to change the name in the "Save as" dialog, to get the unchanged file (hopefully) saved, which you can upload to your board. Again, do not use some sort of editor-style process or copy/paste of file content.

**Reply by Sum-Guy** [reply]
When I right-click on the link you posted, and select "save link as", my browser (firefox 102.4) offers to save a file called ST7789_display.ini.  That's great.  I save it.  The file is 216 kb and it's the entire html code for that page.  For some reason I'm not able to actually download just that file.  I ordinarily don't have problems downloading files with browsers.  Github is sending me the entire web page when I request just that .ini file.

Unbelievable.   I try 

wget https://github.com/arendst/Tasmota/blob/development/tasmota/displaydesc/ST7789_display.ini

and still I get a 151 kb file containing web-code.

I told you it isin't that easy to download repository files from github.  I've never figured out how to do it.

**Reply by sfromis** [reply]
Do not right-click on my link, instead do specifically as I suggested.

Tested with Firefox, on the specific page I linked, **right-clicking exactly on the "Raw" button** and "Save Link As" gave me the option to save the file, resulting in a file of the expected size of 227 bytes. 
![image](https://user-images.githubusercontent.com/47082390/210288071-b5f3ccc8-b5fb-4e2a-a4a1-8ab7d493caae.png)

Your wget test would've been ok if you did not point at the html page. Clicking on the "Raw" button gave me this URL which should work with wget:
https://raw.githubusercontent.com/arendst/Tasmota/development/tasmota/displaydesc/ST7789_display.ini

**Reply by Sum-Guy** [reply]
It might have been more helpful to have posted the link to the file in the first place (the raw.what-ever URL).

I've downloaded that file.  It is 227 bytes.  I downloaded the display.ini that was in the tasmota file system (call this the "bad" file) to my PC.  It is 230 bytes.  I've loaded both into notepad++.  They look identical, except that the "real" file is rendered with a small indentation to the right, this indented zone is colored grey.  It doesn't seem to be a tab, I don't know what it is.

I'm going to analyze the files byte by byte to see how they are different.

In the mean time, I did upload the "real" file to tasmota, and bingo, something is now working.

The date and time are being displayed in very tiny characters, but in irritating way, it is flashing about once per second.  Does it have to flash?

I will look at the files and post later here what the differences are.  Just to say again, I used "create new file" method by pasting the text directly in the tasmota GUI to create the "bad" display.ini.  

-----------

displaymode 2 - the device prints out a bunch of info, wifi info, IP address settings.   It's supposed to print out local sensor data.  which I see that it does - after a few minutes....?

When you change displaymode, is it supposed to scroll through a bunch of device config info (and then freezes for a few minutes displaying wifi SSID and signal strength) before it finally displays what it's supposed to display?

I have it now set to displaymode 3, it's displaying temp and humidity.  It's also supposed to be displaying date and time in this mode (it's not).

**Reply by sfromis** [reply]
I made the choice of telling you how to download, since you had trouble doing that. But you ignored the instructions, and did something entirely different.

**Comment by Sum-Guy** [comment]
I'm going to have to try this again, to manually create the display.ini by using the tasmota gui create new file, but this is what I've found.

The ST7789_display.ini file is 227 bytes.

The display.ini file that was in the tasmota file system and was created by cutting and pasting the text from the above ini file was 230 bytes after I downloaded it from tasmota to my PC.

The extra 3 bytes is the sequence 239 187 191 at the start of the file.  All other bytes are identical.

So what are those 3 bytes?  Well, turns out they are called the "byte order mark".

It's the utf8 byte order mark. Some editors save the BOM inside the file (in order to be used as a header) which regularly causes confusion because it is optional.

The byte order mark (BOM) is a Unicode character, U+FEFF BYTE ORDER MARK (BOM), whose appearance as a magic number at the start of a text stream can signal several things to a program reading the text:

The byte order, or endianness, of the text stream;
The fact that the text stream's encoding is Unicode, to a high level of confidence;
Which Unicode encoding the text stream is encoded as.

See also:

http://www.javapractices.com/topic/TopicAction.do?Id=257

**Reply by sfromis** [reply]
Byte order mark and such is the reason why default Windows text editors are **bad** to use for raw text files, and cut/paste is no defense against this. You can find a better editor, suitable for raw text, instead. For such files, Unicode is not relevant to use, but instead "old-fashioned" 7-bit Ascii for safe text, where endianness is not an issue. Another reason for Windows tools routinely messing up text files is insisting on line breaks being cr+lf, instead of the otherwise most common lf only.

Anyway, you do not *need* to manually create the file, if you just download it the right way, as I *again* described to you.
